### PR TITLE
PR for expe on apcg

### PR DESCRIPTION
**Pull request by QB3**

### CODE REVIEW

File: andersoncd/lasso.py [hunk 1]
```diff
@@ -381,27 +390,34 @@ def apcg(X, y, alpha, max_iter=10000, tol=1e-4, f_gap=10, verbose=False):
...
         else:
-            tau, tau_old = _apcg(X, z, u, tau, Xu, Xz, y, alpha, lc)
+            tau, tau_old = _apcg(X, z, u, tau, Xu, Xz, y, alpha, rho, lc)
 
         if it % f_gap == 0:
             w = tau_old ** 2 * u + z
             R = y - X @ w  # MM: todo this is brutal if f_gap = 1
 
-            p_obj = primal_enet(R, w, alpha)
+            p_obj = primal_enet(R, w, alpha, rho=rho)
```
Comment (mathurinm) [inline andersoncd/lasso.py:402]: just rho (not rho=rho) is lighter IMO

File: andersoncd/lasso.py [hunk 2]
```diff
@@ -381,27 +390,34 @@ def apcg(X, y, alpha, max_iter=10000, tol=1e-4, f_gap=10, verbose=False):
...
             if np.abs(p_obj) > np.abs(E[0] * 1e3):
                 break
 
             if alpha != 0:
                 theta = R / alpha
-                d_norm_theta = np.max(np.abs(X.T @ theta))
-                if d_norm_theta > 1.:
-                    theta /= d_norm_theta
-                d_obj = dual_lasso(y, theta, alpha)
+                # d_norm_theta = np.max(np.abs(X.T @ theta))
```
Comment (mathurinm) [inline andersoncd/lasso.py:410]: remove commented code

File: andersoncd/tests/test_enet.py [hunk 3]
```diff
@@ -28,12 +29,19 @@ def test_enet_solver(algo, use_acc, l1_ratio):
...
+            X, y, alpha * l1_ratio * n_samples,
+            rho=alpha * (1 - l1_ratio) * n_samples, tol=tol, verbose=True,
+            max_iter=100_000)[0]
+    else:
+        coef_ours = solver_enet(
+            X, y, alpha * l1_ratio * n_samples,
+            rho=alpha * (1 - l1_ratio) * n_samples, tol=tol, algo=algo,
+            use_acc=use_acc, max_iter=1_000_000)[0]
+
+    print(coef_ours - coef_sk)
```
Comment (mathurinm) [inline andersoncd/tests/test_enet.py:43]: remove print

File: examples/expe_conv_enet/figure_enet.py [hunk 4]
```diff
@@ -149,7 +150,23 @@
     axarr_E.flat[idx1 * len(div_alphas)].set_ylabel(
         r"$\rho = \lambda / %i$" % div_rho)
 
-fig.show()
-fig_E.show()
+save_fig = False
+# save_fig = True
+fig_dir = "../../../extrapol_cd/tex/aistats20/prebuiltimages/"
```
Comment (mathurinm) [inline examples/expe_conv_enet/figure_enet.py:155]: remove

File: examples/expe_conv_lasso/figure_lasso.py [hunk 5]
```diff
@@ -3,21 +3,25 @@
...
 # div_alphas = [10, 100, 1000, 5000]
 
 
 # if you want to run the expe quickly choose instead:
-dataset_names = [
-    "leukemia", "gina_agnostic", ]
-div_alphas = [10, 100]
-
-################## config #####################################################
+# dataset_names = ["hiva_agnostic", 'rcv1_train']
```
Comment (mathurinm) [inline examples/expe_conv_lasso/figure_lasso.py:16]: cleanup

File: examples/expe_conv_lasso/figure_lasso.py [hunk 6]
```diff
@@ -161,7 +165,23 @@
     axarr_E.flat[idx1 * len(div_alphas)].set_ylabel(
         "%s" % dataset_title[dataset])
 
-fig.show()
-fig_E.show()
+
+save_fig = False
```
Comment (mathurinm) [inline examples/expe_conv_lasso/figure_lasso.py:169]: cleanup

File: examples/expe_conv_lasso/main_lasso.py [hunk 7]
```diff
@@ -10,15 +10,17 @@
...
 
 
 # if you want to run the file quickly choose instead:
-dataset_names = [
-    "leukemia", "gina_agnostic", ]
-div_alphas = [10, 100]
+# dataset_names = ["rcv1_train"]
+div_alphas = [5_000]
+
+# div_alphas = [1000]
```
Comment (mathurinm) [inline examples/expe_conv_lasso/main_lasso.py:22]: cleanup

File: examples/expe_conv_logreg/figure_logreg.py [hunk 8]
```diff
@@ -123,6 +131,23 @@
     axarr_E.flat[idx1 * len(div_alphas)].set_ylabel(
         "%s" % dataset_title[dataset])
 
+save_fig = False
```
Comment (mathurinm) [inline examples/expe_conv_logreg/figure_logreg.py:134]: remove all below

File: examples/plot_intro_ols.py [hunk 9]
```diff
@@ -56,11 +54,18 @@ def callback(x):
...
-    ('pgd', True),
-    ('fista', False),
     ('cd', False),
+    # cg here
+    ('pgd', True),
     ('cd', True),
-    ('apcg', False),
+    ('fista', False),
+    ('apcg', False)
+    # ('pgd', False),
```
Comment (mathurinm) [inline examples/plot_intro_ols.py:63]: cleanup

File: andersoncd/plot_utils.py [hunk 10]
```diff
@@ -26,3 +27,17 @@ def configure_plt():
     sns.set_palette('colorblind')
     sns.set_context("poster")
     sns.set_style("ticks")
+
+
+def plot_legend_apart(ax, figname, ncol=None):
```
Comment (mathurinm) [inline andersoncd/plot_utils.py:32]: maybe make this private as Im unsure pdfcrop is installed on every system and it's for our own use only ?

File: examples/expe_conv_enet/figure_enet.py [hunk 11]
```diff
@@ -74,6 +74,7 @@
 
 
 ####################################
+# div_alphas = [10, 100]
```
Comment (mathurinm) [inline examples/expe_conv_enet/figure_enet.py:77]: rm

File: examples/expe_conv_enet/main_enet.py [hunk 12]
```diff
@@ -9,20 +9,28 @@
...
+# dataset_names = [
+#     "leukemia", "gina_agnostic", "hiva_agnostic", 'rcv1_train', 'news20']
 
 
+div_alphas = [10, 100, 1000, 5_000]
+div_rhos = [10, 100]
+
+
+n_jobs = 1
+# n_jobs = len(dataset_names) * len(div_alphas) * len(algos)
```
Comment (mathurinm) [inline examples/expe_conv_enet/main_enet.py:22]: rm

File: examples/expe_conv_lasso/figure_lasso.py [hunk 13]
```diff
@@ -3,21 +3,20 @@
...
-div_alphas = [10, 100]
+    "leukemia", "gina_agnostic", "hiva_agnostic", 'rcv1_train']
+div_alphas = [10, 100, 1_000, 5_000]
 
-################## config #####################################################
+# if you want to run the file quickly choose instead:
+# dataset_names = ["leukemia", "gina_agnostic", "hiva_agnostic", 'rcv1_train']
+# div_alphas = [10, 100]
+
+"""Config
```
Comment (mathurinm) [inline examples/expe_conv_lasso/figure_lasso.py:18]: rm?

File: examples/expe_conv_lasso/main_lasso.py [hunk 14]
```diff
@@ -10,15 +10,18 @@
...
 # if you want to run the file quickly choose instead:
-dataset_names = [
-    "leukemia", "gina_agnostic", ]
-div_alphas = [10, 100]
+# dataset_names = ["leukemia", "gina_agnostic", "hiva_agnostic", 'rcv1_train']
+# div_alphas = [10, 100]
+
+
+n_jobs = 1
+# n_jobs = len(dataset_names) * len(div_alphas) * len(algos)
```
Comment (mathurinm) [inline examples/expe_conv_lasso/main_lasso.py:23]: rm

File: examples/expe_conv_logreg/main_logreg.py [hunk 15]
```diff
@@ -2,29 +2,33 @@
...
-dataset_names = [
-    "gina_agnostic", 'rcv1_train']
-div_alphas = [10]
+# dataset_names = ["gina_agnostic"]
+# div_alphas = [10, 100]
 
 
-############################## config #########################################
+n_jobs = 1
+# n_jobs = len(dataset_names) * len(div_alphas) * len(algos)
```
Comment (mathurinm) [inline examples/expe_conv_logreg/main_logreg.py:22]: rm

File: examples/expe_conv_logreg/main_logreg.py [hunk 16]
```diff
@@ -82,36 +86,39 @@
...
 
 
 def parallel_function(dataset_name, algo, div_alpha):
     algo_name, use_acc, K = algo
     if dataset_name.startswith((
             'rcv1_train', 'news20', 'kdda_train', 'finance')):
-        X, y = fetch_libsvm(dataset_name)
-        X.multiply(1 / sparse.linalg.norm(X, axis=0))
+        X, y = fetch_libsvm(dataset_name, normalize=True)
+        # X.multiply(1 / sparse.linalg.norm(X, axis=0))
```
Comment (mathurinm) [inline examples/expe_conv_logreg/main_logreg.py:99]: rm

File: examples/plot_intro_ols.py [hunk 17]
```diff
@@ -116,21 +114,35 @@ def callback(x):
...
 plt.ylabel(r"$f(x^{(k)}) - f(x^{*})$")
 plt.xlabel(r"iteration $k$")
 ax.set_yticks((1e-15, 1e-10, 1e-5, 1e0))
-plt.title("Convergence on Least Squares")
 plt.tight_layout()
 
 
+save_fig = False
+# save_fig = True
+fig_dir = "../../extrapol_cd/tex/aistats20/prebuiltimages/"
```
Comment (mathurinm) [inline examples/plot_intro_ols.py:135]: update

File: examples/expe_conv_enet/main_enet.py (hunk 12 above)
Comment (QB3) [reply examples/expe_conv_enet/main_enet.py:22]: why? I just set n_jobs to 1 to be sure that someone not experienced with the script won't launch it with 15 jobs, but it seems logical to let it no?

**Pull request merged**